### PR TITLE
add validation for empty telemetry metric address when level not none

### LINKED ISSUE #5655: Setting an empty metrics address should fail fast
I would expect an empty address [`telemetry::metrics::address`] to return a validation error, instead of silently disabling telemetry. Do you think this is worth an issue?

_Originally posted by @codeboten in https://github.com/open-telemetry/opentelemetry-collector/pull/5642#discussion_r916833143_

### PR DESCRIPTION
**Pull request by fatsheep9146**
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

**Description:** 
Fix #5655

### CODE REVIEW

**Review by TylerHelmuth — APPROVED** [review]
LGTM

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 ### 💡 Enhancements 💡
 
 - Add `linux-ppc64le` architecture to cross build tests in CI
+- Add validation for empty address [telemetry::metrics::address] (#5661)
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply CHANGELOG.md:8]: +1. Not only that, but we should do this change in two stages: one deprecating the old behavior, another one changing it on a subsequent version.

[the remark this answers is not in the document]
Comment (jpkrohling) [reply CHANGELOG.md:8]: Can you check the deprecation guidelines from the contribution doc? We should have the process outlined there, along with a few examples.

File: config/moved_config.go [hunk 2]
```diff
@@ -147,6 +148,12 @@ func (cfg *Config) validateService() error {
 			}
 		}
 	}
+
+	// Check when service telemetry metric level is not none, the metrics address should not be empty
+	if cfg.Service.Telemetry.Metrics.Level != configtelemetry.LevelNone && cfg.Service.Telemetry.Metrics.Address == "" {
+		return fmt.Errorf("collector telemetry metric address should exist when metric level is not none")
+	}
+
```
Comment (bogdandrutu) [inline config/moved_config.go:156]: [Idea]: On `cfg.Service.Telemetry` add a `Validation` func and call that one.

What do you think?

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (fatsheep9146) [reply CHANGELOG.md:8]: I've checked the part [Breaking changes](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#breaking-changes) of contributing guideline. 
The listed examples: `Example #1 - Renaming a function`, `Example #2 - Changing the return values of a function` and `Example #3 - Changing the arguments of a function` all do not fit this case.

But I think this case maybe more likely to 
> when renaming or refactoring types, functions, or attributes, we MUST create the new name and MUST deprecate the old one in one version (step 1), and MAY remove it in a subsequent version (step 2). For simple renames, the old name SHALL call the new one.

So I think we should do 2 steps
- in version N+1, keep `validateService` almost unchanged but add a warning log of deprecation, and add `mustValidateService` function to add the new validation logic of telemetry metric
- in version N+2, change `mustValidateService` to become the new version of `validateService`

@jpkrohling @TylerHelmuth

[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply CHANGELOG.md:8]: I believe that follows the cadence correctly.

File: config/moved_config.go (hunk 2 above)
Comment (fatsheep9146) [reply config/moved_config.go:156]: I saw other config package use `Validate`, not `Validation`, I think I better keep up?

@bogdandrutu

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (fatsheep9146) [reply CHANGELOG.md:13]: I already changed code, please help review this again =D @TylerHelmuth @jpkrohling

File: config/moved_config.go (hunk 2 above)
Comment (fatsheep9146) [reply config/moved_config.go]: @bogdandrutu I already changed this with `Validate`

Comment (bogdandrutu) [reply config/moved_config.go]: yes, `Validate` sorry.

**Review by bogdandrutu — COMMENTED** [review]
Please fix lint issues

**Comment by fatsheep9146** [comment]
@bogdandrutu @jpkrohling @TylerHelmuth All checks are passed, please help review this again, thanks, =D

File: config/moved_config.go [hunk 3]
```diff
@@ -96,6 +96,7 @@ func (cfg *Config) Validate() error {
 	return cfg.validateService()
 }
 
+// Deprecated: [v0.55.0] Use mustValidateService instead.
```
Comment (bogdandrutu) [inline config/moved_config.go:99]: This is private, why do you need deprecation?

Comment (fatsheep9146) [reply config/moved_config.go:99]: I didn't notice that. how to indicate it is deprecated for private function? could your give me a hint?

Comment (bogdandrutu) [reply config/moved_config.go:99]: just break it/change it... nobody except us is using that func.

Comment (fatsheep9146) [reply config/moved_config.go]: I've already removed the comment.

**Comment by fatsheep9146** [comment]
@bogdandrutu I think the failed check is not related to this pr.

**Comment by jpkrohling** [comment]
PR rebased.

**Comment by fatsheep9146** [comment]
> PR rebased.

Done

**Comment by jpkrohling** [comment]
The CI is failing due to https://github.com/open-telemetry/opentelemetry-collector/issues/5770

**Comment by fatsheep9146** [comment]
> The CI is failing due to #5770

Done

**Comment by fatsheep9146** [comment]
ping @jpkrohling, could you review this again?

**Comment by jpkrohling** [comment]
`unittest (1.19)` didn't complete yet, PR rebased to see if this check is triggered.

**Comment by fatsheep9146** [comment]
@jpkrohling unittest for 1.19 is triggered and passed =D

**Comment by fatsheep9146** [comment]
ping @jpkrohling for review this again, thx, =D

**Comment by jpkrohling** [comment]
Sorry, I was out last week. Could you please rebase this? There seems to be a conflict that needs solving.

**Comment by fatsheep9146** [comment]
> Sorry, I was out last week. Could you please rebase this? There seems to be a conflict that needs solving.

I already rebased this.

**Comment by fatsheep9146** [comment]
Could you review this again? @jpkrohling

**Comment by jpkrohling** [comment]
@bogdandrutu, you had comments on this before as well, do you want to review it again?

**Comment by fatsheep9146** [comment]
ping @bogdandrutu @jpkrohling

File: config/moved_config.go [hunk 4]
```diff
@@ -147,9 +147,20 @@ func (cfg *Config) validateService() error {
 			}
 		}
 	}
+
 	return nil
 }
 
+// since mustValidateService will be used for next version, so use nolint to escape the `unused` lint error
+// nolint
+func (cfg *Config) mustValidateService() error {
```
Comment (jpkrohling) [inline config/moved_config.go:156]: In Go, "Must*" functions typically mean that the call never fails (no errors), resulting in panic in case of an error. I'm not saying you should panic here, just that you should set the expectation straight by not using "must" as part of the name.

File: config/moved_config.go [hunk 5]
```diff
@@ -147,9 +147,20 @@ func (cfg *Config) validateService() error {
...
+	if err := cfg.validateService(); err != nil {
+		return err
+	}
+
+	return cfg.Telemetry.Validate()
```
Comment (jpkrohling) [inline config/moved_config.go:161]: Shouldn't this be part of the main Validate function? Why do you need this in a specific extra function (mustValidateService)?

File: config/moved_config.go (hunk 4 above)
Comment (fatsheep9146) [reply config/moved_config.go:156]: That make senses. I originally tried to follow [example 2](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#example-2---changing-the-return-values-of-a-function), so I name the new function to mustValidateService.

What name do you think is better, is `newValidateService` or `newValidateService` ok?  @jpkrohling

File: config/moved_config.go (hunk 5 above)
Comment (fatsheep9146) [reply config/moved_config.go:161]: This also follows [example 2](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#example-2---changing-the-return-values-of-a-function), since I introduced a new validating rule for config file, which may result that the collector with wrong config file crashes when upgraded to new version if the pr is merged. 
So I first mark old validation function as deprecated and use the new validation for next release.

Comment (jpkrohling) [reply config/moved_config.go:161]: Oh, I see the confusion now. The example shows how one would proceed in changing the signature of a function to add a new return value (an error).

In your case, you should just add the new condition to the function. But instead of returning an error, you should just log a warning for a couple of releases and then return an error. In this case, given the bug is because the validation does not exist for this condition, adding the condition wouldn't be considered a breaking change (even though it would indeed break already faulty instances).

cc @open-telemetry/collector-contrib-approvers, for more opinions on this.

Comment (fatsheep9146) [reply config/moved_config.go:161]: > Oh, I see the confusion now. The example shows how one would proceed in changing the signature of a function to add a new return value (an error).
> 
> In your case, you should just add the new condition to the function. But instead of returning an error, you should just log a warning for a couple of releases and then return an error. In this case, given the bug is because the validation does not exist for this condition, adding the condition wouldn't be considered a breaking change (even though it would indeed break already faulty instances).
> 
> cc @open-telemetry/collector-contrib-approvers, for more opinions on this.

Ok, I got it. So I need to do following things
- remove `mustValidateService` function, just add new validation to original `validateService` function
- for next release(N+1), just log a warning without return error, so the instance with faulty config can still start.
- for N+2 release, return error for this validation. 

@jpkrohling Do I understand this correctly?

Comment (jpkrohling) [reply config/moved_config.go:161]: Yes, sounds good to me.

Comment (fatsheep9146) [reply config/moved_config.go]: @jpkrohling  I already remove `mustValidateService` and move the validation of Telemetry config directly into `validateService ` function, and in validateService function I initialize a zap.logger to log for the validation fail. Please help review this, thank you =D

File: service/collector_test.go [hunk 6]
```diff
@@ -518,27 +518,36 @@ func TestCollectorShutdownBeforeRun(t *testing.T) {
...
-	require.Error(t, col.Run(context.Background()))
+			// Expect run to error
+			require.Error(t, col.Run(context.Background()))
 
-	// Expect state to be closed
-	assert.Equal(t, Closed, col.GetState())
+			// Expect state to be closed
+			assert.Equal(t, Closed, col.GetState())
+		})
+	}
```
Comment (bogdandrutu) [inline service/collector_test.go:550]: Remove unnecessary/unrelated changes.

File: config/moved_config.go [hunk 7]
```diff
@@ -97,6 +100,31 @@ func (cfg *Config) Validate() error {
...
+		InitialFields:     cfg.Telemetry.Logs.InitialFields,
+	}
+	if zapCfg.Encoding == "console" {
+		// Human-readable timestamps for console format of logs.
+		zapCfg.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
+	}
+	logger, err := zapCfg.Build()
+	if err != nil {
+		return err
+	}
```
Comment (bogdandrutu) [inline config/moved_config.go:126]: Why this change?

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
Lots of unrelated changes, please keep only changes in service/telemetry

File: config/moved_config.go (hunk 7 above)
Comment (fatsheep9146) [reply config/moved_config.go:126]: > > Oh, I see the confusion now. The example shows how one would proceed in changing the signature of a function to add a new return value (an error).
> > In your case, you should just add the new condition to the function. But instead of returning an error, you should just log a warning for a couple of releases and then return an error. In this case, given the bug is because the validation does not exist for this condition, adding the condition wouldn't be considered a breaking change (even though it would indeed break already faulty instances).
> > cc @open-telemetry/collector-contrib-approvers, for more opinions on this.
> 
> Ok, I got it. So I need to do following things
> 
> * remove `mustValidateService` function, just add new validation to original `validateService` function
> * for next release(N+1), just log a warning without return error, so the instance with faulty config can still start.
> * for N+2 release, return error for this validation.
> 


Becase I talked about the with @jpkrohling , we should first log with warning for next release(N+1), and return error in N+2 release. So I have to output a warning log in validateService function of config/move_config.go. 

But for now, the config.Config does not have a logger field in it. In order to keep the same logging format as collector, so I initialize a logger object same as https://github.com/open-telemetry/opentelemetry-collector/blob/e98116922cf6c079c246fa78eef59dfd9c2e4867/service/internal/telemetrylogs/logger.go#L26

I also think this is not the right way to output warning log in config/move_config.go, do you have any suggestion for this? @bogdandrutu

Comment (jpkrohling) [reply config/moved_config.go:126]: It does; would that be a bit too much though? Not sure I see users wanting to enable this feature gate themselves.

Comment (fatsheep9146) [reply config/moved_config.go:126]: Honestly, I prefer the logging way, since the validation of config may be enhanced again in the future, not only in this pull request, warning log may be more friendly to both developers and users.

@bogdandrutu @jpkrohling

Comment (bogdandrutu) [reply config/moved_config.go:126]: Users will start to ask to configure that, etc. If you want to log something then simply return an error with a good message. Also why configuring a `zap.Logger` vs `fmt.Printf`?

Comment (fatsheep9146) [reply config/moved_config.go:126]: Honestly, how to log the warning message also puzzled me for a while. Finally I choose `zap.Logger` because I want to keep the same logging experience with other logs for collector, as following (the first line):

```
2022-09-29T09:17:18.442+0800	warn	config/moved_config.go:180	Telemetry config validation failed	{"reason": "collector telemetry metric address should exist when metric level is not none"}

2022-09-29T09:17:18.442+0800	info	service/telemetry.go:107	Skipping telemetry setup.	{"address": "", "level": "basic"}
2022-09-29T09:17:18.442+0800	info	components/components.go:30	In development component. May change in the future.	{"kind": "exporter", "data_type": "traces", "name": "logging", "stability": "in development"}
```

If use `fmt.Printf` instead, the experience will be
```
Telemetry config validation failed: collector telemetry metric address should exist when metric level is not none

2022-09-29T09:22:57.747+0800	info	service/telemetry.go:107	Skipping telemetry setup.	{"address": "", "level": "basic"}
2022-09-29T09:22:57.747+0800	info	components/components.go:30	In development component. May change in the future.	{"kind": "exporter", "data_type": "traces", "name": "logging", "stability": "in development"}
```

I'm not saying keep the uniform logging experience is necessary, I think we can talk more about this and agree on a right way.

@bogdandrutu

Comment (fatsheep9146) [reply config/moved_config.go:126]: Do you have more suggestions? @bogdandrutu

Comment (fatsheep9146) [reply config/moved_config.go:126]: ping @bogdandrutu @jpkrohling 

Do you have any suggestion for push forwarding this pr?

Comment (jpkrohling) [reply config/moved_config.go:126]: I still think that we should log a warning message in case we have an empty telemetry metrics address and level is not none, returning an error in a future version. I don't think this deserves to be behind a feature gate, though. But I'm good with whatever @bogdandrutu decides.

Comment (fatsheep9146) [reply config/moved_config.go:126]: > Honestly, how to log the warning message also puzzled me for a while. Finally I choose `zap.Logger` because I want to keep the same logging experience with other logs for collector, as following (the first line):
> 
> ```
> 2022-09-29T09:17:18.442+0800	warn	config/moved_config.go:180	Telemetry config validation failed	{"reason": "collector telemetry metric address should exist when metric level is not none"}
> 
> 2022-09-29T09:17:18.442+0800	info	service/telemetry.go:107	Skipping telemetry setup.	{"address": "", "level": "basic"}
> 2022-09-29T09:17:18.442+0800	info	components/components.go:30	In development component. May change in the future.	{"kind": "exporter", "data_type": "traces", "name": "logging", "stability": "in development"}
> ```
> 
> If use `fmt.Printf` instead, the experience will be
> 
> ```
> Telemetry config validation failed: collector telemetry metric address should exist when metric level is not none
> 
> 2022-09-29T09:22:57.747+0800	info	service/telemetry.go:107	Skipping telemetry setup.	{"address": "", "level": "basic"}
> 2022-09-29T09:22:57.747+0800	info	components/components.go:30	In development component. May change in the future.	{"kind": "exporter", "data_type": "traces", "name": "logging", "stability": "in development"}
> ```
> 
> I'm not saying keep the uniform logging experience is necessary, I think we can talk more about this and agree on a right way.
> 
> @bogdandrutu

@bogdandrutu Do you have more suggestions? 
I think we can all agree with the opinion that we should use warning logs, not feature-gates.
Since this package does not have import `zap.Logger` , should we just use fmt.Printf?

Comment (bogdandrutu) [reply config/moved_config.go:126]: Yes, let's use `fmt.Printf` then

Comment (fatsheep9146) [reply config/moved_config.go]: Could you help review this again? @bogdandrutu

Comment (fatsheep9146) [reply config/moved_config.go]: Ping @bogdandrutu

**Review by bogdandrutu — APPROVED** [review]
